src/Build/BackEnd/Components/Communications/NodeProviderOutOfProcBase.cs

```diff
@@ -480,8 +479,8 @@ private Process LaunchNode(string msbuildLocation, string commandLineArgs)
             if (!NativeMethodsShared.IsMono)
             {
                 // Run the child process with the same host as the currently-running process.
-                exeName = GetCurrentHost();
-                commandLineArgs = "\"" + msbuildLocation + "\" " + commandLineArgs;
+                string dotnetExe = Path.Combine(FileUtilities.GetFolderAbove(exeName, 2), "dotnet.exe");
```

Comment:
There is a lot going on in this method:

* On Windows, it uses native methods to launch the process so the path to `MSBuild.exe` needs to be prepended to the command-line arguments.  @rainersigwald do you know why we use a different method to launch the process?  The code doesn't seem to need to as far as I can tell.  Maybe we can get rid of it?
* For .NET Core, the command-line arguments must be prepended with the the path to `MSBuild.dll` and the EXE is `dotnet`.

I see some missing logic in the current implementation:

* On Windows, its `dotnet.exe`, on non-Windows its just `dotnet`
* The path to `MSBuild.dll` is prepended twice on Windows (which technically works but sees dirty)
* `GetCurrentHost()` is only ever correct if you run `dotnet` directly so using it as a fallback seems not great
* 

I would do the following:

1. Make a method like `GetNodeExe()` that has all of the logic to find the path to the executable that represents the node, so `MSBuild.exe` on .NET Framework, `dotnet.exe` or `dotnet` for .NET Core.
2. Make a method like `GetNodeArguments()` that adjusts the command-line arguments
  Prepend `MSBuild.exe` on Windows since its using native APIs
  Prepend `MSBuild.dll` on .NET Core

In my opinion that would simplify this method and make it easier to rationalize what its doing.  You could also make these two methods `internal` and write a unit test to verify they return the correct values.

Reply:
As far as falling back to GetCurrentHost, I agree it isn't terribly likely to succeed, but I don't think we have a better option. There's a chance someone has fancy logic in their project to make it work, and if we don't fall back to GetCurrentHost, we would break that.

Reply:
I'm pretty sure if you don't fall back to current-host our own dogfood builds will fail (they're invoked via `dotnet.exe path\to\some\msbuild.dll`).

Reply:
> GetCurrentHost() is only ever correct if you run dotnet directly so using it as a fallback seems not great

Sorry what I meant was having this method here only to use as a fallback, when the entirety of the logic could live inside of it.  I did not mean to get rid of the fallback all together.  To have a solid logic here that then falls back to a method seems clunky is all.